python/ray/autoscaler/_private/event_system.py

```diff
@@ -0,0 +1,97 @@
+        acquiring_new_head_node : Invoked before the head node is acquired.
+        head_node_acquired : Invoked after the head node is acquired.
+        ssh_control_acquired : Invoked when the node is being updated.
+        run_initialization_cmd : Invoked before all initialization
+            commands are called and again before each initialization command.
+        run_setup_cmd : Invoked before all setup commands are
+            called and again before each setup command.
+        start_ray_runtime : Invoked before ray start commands are run.
+        start_ray_runtime_completed : Invoked after ray start commands
+            are run.
+        cluster_booting_completed : Invoked after cluster booting
+            is completed.
+    """
+
+    up_started = auto()
+    ssh_keypair_downloaded = auto()
+    cluster_booting_started = auto()
+    acquiring_new_head_node = auto()
+    head_node_acquired = auto()
+    ssh_control_acquired = auto()
+    run_initialization_cmd = auto()
+    run_setup_cmd = auto()
+    start_ray_runtime = auto()
+    start_ray_runtime_completed = auto()
+    cluster_booting_completed = auto()
+
+
+class _EventSystem:
+    """Event system that handles storing and calling callbacks for events.
+
+    Attributes:
+        callback_map (Dict[str, List[Callable]]) : Stores list of callbacks
+            for events when registered.
+    """
+
+    def __init__(self):
+        self.callback_map = {}
+
+    def add_callback_handler(
+            self,
+            event: str,
+            callback: Union[Callable[[Dict], None], List[Callable[[Dict],
+                                                                  None]]],
+    ):
+        """Stores callback handler for event.
+
+        Args:
+            event (str): Event that callback should be called on. See
+                CreateClusterEvent for details on the events available to be
+                registered against.
+            callback (Callable[[Dict], None]): Callable object that is invoked
+                when specified event occurs.
+        """
+        if type(callback) is not list:
+            callback = [callback]
+        if event in self.callback_map:
+            self.callback_map[event].extend(callback)
+        else:
+            self.callback_map[event] = callback
+
+    def execute_callback(self, event: str, event_data: Dict[str, Any] = {}):
+        """Executes all callbacks for event.
+
+        Args:
+            event (str): Event that is invoked. See CreateClusterEvent
+                for details on the available events.
+            event_data (Dict[str, Any]): Argument that is passed to each
+                callable object stored for this particular event.
+        """
+        event_data["event_name"] = event
+        if event in self.callback_map:
+            for callback in self.callback_map[event]:
+                callback(event_data)
+
+    def clear_callbacks_for_event(self, event: str):
```

Comment:
Hmm, is this called anywhere?

Reply:
Currently this isn't called anywhere, but it might be useful when we want different callbacks for the same open source function when called from different CLI commands.

Reply:
SGTM!